pallets/xcmp-handler/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # XCMP Handler pallet
+//!
+//! This pallet is used to send XCM messages to other chains.
+//! In order to do that it needs to keep track of what tokens other chains accept,
+//! and the relevant rates.
+//!
+//! At this moment we only support using our native currency. We are looking into supporting
+//! other chain's native currency and then any currency.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod weights;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::Decode;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::{StorageDoubleMap, *},
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_std::prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		///The currencyIds that our chain supports.
+		type CurrencyId: Parameter
+			+ Member
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Ord
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		///The currencyId for the native currency.
+		#[pallet::constant]
+		type GetNativeCurrencyId: Get<Self::CurrencyId>;
+
+		/// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T> {
+		/// XCM data was added for a chain/currency pair.
+		XcmDataAdded { para_id: u32 },
+		/// XCM data was removed for a chain/currency pair.
+		XcmDataRemoved { para_id: u32 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		///We only support certain currency/chain combinations.
+		CurrencyChainComboNotSupported,
+	}
+
+	/// Stores all data needed to send an XCM message for chain/currency pair.
+	#[derive(Clone, Debug, Encode, Decode, PartialEq, TypeInfo)]
+	pub struct XcmCurrencyData {
+		pub native: bool,
+		pub fee_per_second: u128,
+		pub instruction_weight: u64,
+	}
+
+	/// Stores XCM data for a chain/currency pair.
+	#[pallet::storage]
+	#[pallet::getter(fn get_xcm_chain_data)]
+	pub type XcmChainCurrencyData<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, u32, Twox64Concat, T::CurrencyId, XcmCurrencyData>;
```

Comment:
Could you add a type alias for the `u32`? Maybe `ParaId` (unless that name is already overloaded)? It's nice to be explicit about what's being stored.